Demand per day billing determinant

jrogers97 opened the pull request:
Adds a 'demand per day' billing determinant - where the monthly determinant is the sum of the max demand for every day in that month. New test data is taken from [this sheet](https://docs.google.com/spreadsheets/d/1YbKH6BIhUxi0pLZ2jE74lTX0v9K4jDDLSsXqvtiv_ec/edit#gid=1313587342).

jrogers97 commented on src/rateEngine/LoadProfile.ts:
```diff
@@ -97,6 +99,20 @@ class LoadProfile {
     });
   }
 
+  maxPerDayByMonth(): Array<number> {
```
I'm open to a better name for this method if one stands out @bknoles @euggra

euggra replied:
That's a great name.

bknoles replied:
I see why this gives you pause... it's missing the "sum" concept. something like `maxPerDaySummedByMonth` is more accurate. That's a pretty awkward name, which makes me think there's probably another class we could add.

Something that would yield an API like:

```
loadProfile.max(by: 'day').sum(by: 'month');
// or
loadProfile.maxByDay().sumByMonth();
```

That's out of the scope of this PR, but i would be in favor of using the awkward name until we can take a longer look: `maxPerDaySummedByMonth`.

jrogers97 commented on src/rateEngine/LoadProfile.ts:
```diff
@@ -97,6 +99,20 @@ class LoadProfile {
+      const monthLoads = expanded.filter(({ month }) => m === month);
+
+      // chunk monthly loads by day (31-element array for January, etc.)
+      const dayLoads = Object.values(groupBy(monthLoads, (val) => val.date));
```
This was an interesting problem. I originally was manually chunking every month into 24-element arrays but forgot that DST adds some complexity. This solution partitions the month by each reading's `date` field, which is definitely safer

euggra replied:
Good eye!

bknoles commented on src/rateEngine/BillingDeterminantFactory.ts:
```diff
@@ -11,25 +11,28 @@ import SurchargeAsPercent, { SurchargeAsPercentArgs } from './billingDeterminant
-  'HourlyEnergy' |
-  'DemandTimeOfUse' |
-  'DemandTiersInMonths';
+  | 'EnergyTimeOfUse'
```
looks like we need to set up linting/prettier here! (def not necessary for this PR)

bknoles commented on src/rateEngine/BillingDeterminantFactory.ts:
```diff
@@ -60,6 +63,8 @@ class BillingDeterminantFactory {
         return new DemandTiersInMonths(args as BlockedTiersArgs, loadProfile);
       case 'DemandTimeOfUse':
         return new DemandTimeOfUse(args as DemandTimeOfUseArgs, loadProfile);
+      case 'DemandPerDay':
```
good job remembering to add this to the factory! without naming names (cough... me...cough), we've definitely forgotten that a few times

bknoles commented on src/rateEngine/utils/assumptions.ts:
```diff
@@ -0,0 +1 @@
+export const DAYS_PER_MONTH = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
```
as you mentioned above, this is actually not accurate for leap years... should we make this a function that accepts a year? and do we actually have an existing leap year bug in the billing determinants that use this?

bknoles reviewed (changes requested):
Can you rename the load profile method?

If you think it would be easy to account for DST in our DAYS_PER_MONTH, then that too. If it's not trivial we'll just merge without it.

bknoles commented:
actually, jake, i wonder if it would be best to put the `maxPerDaySummedByMonth` logic directly in the billing determinant instead of in the load profile.

i love that you added it to the load profile... that's the right instinct, and to me it really calls out the need for a different abstraction that we don't currently have.

so rather than committing to a long public method name we don't like, we could "hide" the logic in the billing determinant and only add it as a public method on the load profile once we have an API that we feel good about

jrogers97 commented:
@bknoles I moved the max demand by day logic into the billing determinant, and adjusted `DAYS_PER_MONTH` to be a function that takes a year (and assumes not a leap year if one isn't provided, since one billing determinant doesn't have a load profile to reference).

I also added a quick test for daysPerMonth, which might be unnecessary but I figured it doesn't hurt to have

bknoles reviewed (approved):
looks great!!
